File: src/main/scala/com/juanoff/scala/model/DataStructure.scala
package com.juanoff.scala.model

import com.juanoff.scala.types.Comparator

trait DataStructure {
  def add(value: Any): Unit

  def get(index: Int): Any

  def insert(index: Int, value: Any): Unit

  def remove(index: Int): Unit

  def size(): Int
  
  def forEach(action: DoWith): Unit

  def firstThat(predicate: TestIt): Any

  def sort(comp: Comparator): Unit

  def balance(): DataStructure

  def clear(): Unit
}

File: src/main/kotlin/com/juanoff/kotlin/model/DataStructure.kt
package com.juanoff.kotlin.model

import com.juanoff.kotlin.types.Comparator

interface DataStructure {
    fun add(value: Any)

    fun get(index: Int): Any

    fun insert(index: Int, value: Any)

    fun remove(index: Int)

    fun sort(comp: Comparator)

    fun forEach(action: DoWith)

    fun firstThat(predicate: TestIt): Any?

    fun balance(): DataStructure

    fun size(): Int

    fun clear()
}